File: 0146-lru-cache/0146-lru-cache.java
class LRUCache {

    private LRU lru;

    public LRUCache(int capacity) {
        this.lru = new LRU(capacity);
    }
    
    public int get(int key) {
        int x = lru.get(key);
        // System.out.println("get -> "+key);
        // this.lru.print();
        return x;
    }
    
    public void put(int key, int value) {
        this.lru.put(key, value);
        // System.out.println("put -> "+key+" : "+value);
        // this.lru.print();
    }
}

class LRU
{
    private ListNode head;
    private ListNode tail;

    private Map<Integer, ListNode> map;

    private int cap;

    LRU(int cap)
    {
        this.cap = cap;
        this.map = new HashMap<>();
        this.head = null;
        this.tail = this.head;
    }

    int get(int key)
    {
        if (this.map.containsKey(key))
        {
            int val = this.map.get(key).val;
            remove(key);
            put(key, val);
            return val;
        }
        else
        {
            return -1;
        }
    }

    void put(int key, int val)
    {
        //already present, delete it
        remove(key);
        ListNode node = new ListNode(key, val);
        addToHead(node);
        this.map.put(key, node);
        if (this.map.size()>this.cap)
        {
            deleteTail();
        }

    }

    void deleteTail()
    {
        if (this.tail!=null)
        {
            int key = this.tail.key;
            this.map.remove(key);
            if (this.tail.prev!=null)
            {
                ListNode prev = this.tail.prev;
                prev.next = null;
                this.tail = prev;
            }
            else
            {
                this.tail = null;
                this.head = this.tail;
            }
        }
    }

    void addToHead(ListNode node)
    {
        if (this.head==null)
        {
            this.head = node;
            this.tail = node;
            return;
        }
        node.next = this.head;
        this.head.prev = node;
        this.head = node;
    }

    void remove(int key)
    {
        if (this.map.containsKey(key))
        {
            ListNode node = this.map.get(key);
            ListNode prev = node.prev;
            ListNode next = node.next;
            if (node.prev!=null)
            {
                node.prev.next = next;
            }
            if (node.next!=null)
            {
                node.next.prev = prev;
            }
            if (node==this.head)
            {
                this.head = next;
            }
            if (node==this.tail)
            {
                this.tail = prev;
            }
            this.map.remove(key);
        }
    }

    void print()
    {
        ListNode curr = this.head;
        while(curr!=null)
        {
            System.out.print(curr.val+"->");
            curr = curr.next;
        }
        System.out.println();

        curr = this.tail;
        while(curr!=null)
        {
            System.out.print(curr.val+"<-");
            curr = curr.prev;
        }
        System.out.println();
    }

}

class ListNode
{
    int key;
    int val;
    ListNode next;
    ListNode prev;

    ListNode(int key, int val)
    {
        this.key = key;
        this.val = val;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.py
class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.next = None
        self.prev = None

class LL:
    def __init__(self):
        self.head = None
        self.tail = None

    def add_to_right(self, node: Node):
        if self.tail is None:
            self.head = node
            self.tail = node
            return
        self.tail.next = node
        node.prev = self.tail
        self.tail = node

    def remove_at_left(self) -> int:
        node = self.head
        self.head = self.head.next
        self.head.prev = None
        return node.key
    
    def remove_node(self, node: Node):
        left = node.prev
        right = node.next
        if left is not None:
            left.next = node.next
        if right is not None:
            right.prev = node.prev
        if node is self.head:
            self.head = self.head.next
        if node is self.tail:
            self.tail = self.tail.prev
        node.next = None
        node.prev = None
        

class LRUCache:
    
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.count = 0
        self.ll = LL()
        self.key_to_node = {}
        

    def get(self, key: int) -> int:
        if key not in self.key_to_node:
            return -1
        else:
            node = self.key_to_node[key]
            self.ll.remove_node(node)
        self.ll.add_to_right(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if key in self.key_to_node:
            node = self.key_to_node[key]
            self.ll.remove_node(node)
        else:
            self.count += 1
        node = Node(key, value)
        self.ll.add_to_right(node)
        self.key_to_node[key] = node
        if self.count > self.capacity:
            to_remove = self.ll.remove_at_left()
            self.key_to_node.pop(to_remove, None)
            self.count -= 1
        
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)